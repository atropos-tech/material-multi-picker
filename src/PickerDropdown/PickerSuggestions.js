import React from "react";
import { MenuItem, withStyles } from "@material-ui/core";
import { LOADING, NOT_ENOUGH_CHARACTERS, isError, suggestionsPropType } from "../utils";
import { string, func, number, any, object } from "prop-types";
import DefaultError from "./DefaultError";
import DefaultSuggestion from "./DefaultSuggestion";
import DefaultEmptyMessage from "./DefaultEmptyMessage";
import DefaultLoadingMessage from "./DefaultLoadingMessage";
import DefaultMoreCharactersMessage from "./DefaultMoreCharactersMessage";

const HIGHLIGHT_GREY_SHADE = 300;

const styles = theme => ({
    highlighted: {
        backgroundColor: theme.palette.grey[HIGHLIGHT_GREY_SHADE]
    }
});

function PickerSuggestions(
    { suggestions, getItemProps, highlightedIndex, itemToString, inputValue, SuggestionComponent = DefaultSuggestion, ErrorComponent = DefaultError, classes }
) {
    if ( isError(suggestions) ) {
        return (<ErrorComponent error={ suggestions } inputValue={ inputValue } />);
    }
    if ( suggestions === NOT_ENOUGH_CHARACTERS ) {
        return <DefaultMoreCharactersMessage />;
    }
    if ( suggestions === LOADING ) {
        return <DefaultLoadingMessage inputValue={ inputValue } />;
    }
    if ( Array.isArray(suggestions) ) {
        if ( suggestions.length ) {
            return (
                <>
                    {
                        suggestions.map((item, index) => {
                            const itemId = itemToString(item);
                            const isHighlighted = highlightedIndex === index;

                            const menuItemProps = getItemProps({
                                index,
                                item,
                                role: "menuitem",
                                className: isHighlighted ? classes.highlighted : "",
                                style: {
                                    padding: 0,
                                    height: "auto"
                                },
                            });
                            return (
                                <MenuItem className="suggestion" key={ itemId } {...menuItemProps}>
                                    <SuggestionComponent itemId={ itemId } item={ item } isHighlighted={ isHighlighted } inputValue={ inputValue } />
                                </MenuItem>
                            );
                        })
                    }
                </>
            );
        }
        if ( inputValue.length ) {
            return <DefaultEmptyMessage inputValue={ inputValue } />;
        } else {
            return false;
        }
    }

    console.error(suggestions);
    throw new Error("should never happen!");
}

PickerSuggestions.propTypes = {
    suggestions: suggestionsPropType,
    getItemProps: func.isRequired,
    highlightedIndex: number,
    itemToString: func.isRequired,
    inputValue: string.isRequired,
    SuggestionComponent: any,
    ErrorComponent: any,
    classes: object
};

export default withStyles(styles)(PickerSuggestions);
