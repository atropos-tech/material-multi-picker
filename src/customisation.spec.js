/* eslint-env jest */
/* eslint-disable no-magic-numbers */

import React from "react";
import { mount } from "enzyme";
import { act } from "react-dom/test-utils";
import { resetIdCounter } from "downshift";
import MultiPicker from "./index";
import { Chip, Paper, Avatar, Popover, TextField, InputLabel } from "@material-ui/core";
import JssProvider from "react-jss/lib/JssProvider";
import PickerSuggestions from "./PickerDropdown/PickerSuggestions";

const { stringContaining } = expect;

// workaround for non-stable classnames generated in JSS
// https://github.com/mui-org/material-ui/issues/9492#issuecomment-368205258
//
// forces deterministic classnames that may clash in a wider app (but are safe in unit tests)
//
// Should be addressed in future material-ui release, see
//   https://github.com/mui-org/material-ui/issues/14357
function mountStable(reactElement, ...other) {
    const generateClassName = (rule, styleSheet) => `${styleSheet.options.classNamePrefix}-${rule.key}`;
    const elementWithStableClassNames = (
        <JssProvider generateClassName={ generateClassName }>
            { reactElement }
        </JssProvider>
    );
    return mount(elementWithStableClassNames, ...other);
}

const NOOP = () => { /* do nothing */ };
const SHORT_DELAY_MILLISECONDS = 30;

function delay(delayInMilliseconds = SHORT_DELAY_MILLISECONDS) {
    return new Promise(resolve => {
        setTimeout(resolve, delayInMilliseconds);
    });
}

async function changeInputValueAndUpdate(wrapper, newInputValue) {
    wrapper.find("input").simulate("change", { target: { value: newInputValue}});
    await delay(100);
    wrapper.update();
}

describe("MultiPicker component", () => {

    // avoid non-deterministic id creation for downshift subcomponents
    // breaking snapshot testing
    beforeEach(resetIdCounter);

    it("renders custom chip labels", () => {
        expect.assertions(3);

        const props = {
            itemToString: item => item,
            itemToLabel: jest.fn(() => "some label"),
            value: ["some-item"],
            onChange: NOOP,
            getSuggestedItems: () => []
        };
        const wrapper = mountStable(<MultiPicker { ...props } />);
        expect(wrapper).toContainExactlyOneMatchingElement(Chip);
        expect(wrapper.find(Chip).getDOMNode()).toMatchSnapshot();

        expect(props.itemToLabel).toHaveBeenCalledWith("some-item");
    });

    it("renders custom chip avatars", () => {
        expect.assertions(3);

        const props = {
            itemToString: item => item,
            itemToAvatar: jest.fn(() => <Avatar src='./missing-image' />),
            value: ["some-item"],
            onChange: NOOP,
            getSuggestedItems: () => []
        };
        const wrapper = mountStable(<MultiPicker { ...props } />);
        expect(wrapper).toContainExactlyOneMatchingElement(Avatar);
        expect(wrapper.find(Avatar).getDOMNode()).toMatchSnapshot();

        expect(props.itemToAvatar).toHaveBeenCalledWith("some-item");
    });

    it("renders custom chip colors", () => {
        expect.assertions(1);

        const props = {
            itemToString: item => item,
            chipColor: "secondary",
            value: ["some-item"],
            onChange: NOOP,
            getSuggestedItems: () => []
        };
        const wrapper = mountStable(<MultiPicker { ...props } />);
        expect(wrapper.find(Chip)).toHaveProp("color", "secondary");
    });

    it("renders custom suggestion components", async () => {
        expect.assertions(3);

        function CustomSuggestion() {
            return <span>Some Custom Suggestion</span>;
        }

        const props = {
            itemToString: item => item,
            value: [],
            onChange: NOOP,
            getSuggestedItems: () => ["some suggestion"],
            SuggestionComponent: jest.fn(CustomSuggestion)
        };
        const wrapper = mountStable(<MultiPicker {...props }/>);
        expect(wrapper).not.toContainMatchingElement(Paper);

        await changeInputValueAndUpdate(wrapper, "some text");

        expect(props.SuggestionComponent).toHaveBeenCalledWith({
            itemId: "some suggestion",
            item: "some suggestion",
            isHighlighted: false,
            inputValue: "some text"
        }, {});

        expect(wrapper.find(PickerSuggestions).getDOMNode()).toMatchSnapshot();
    });

    it("shows an custom error message if the user provides the ErrorComponent prop", async () => {
        expect.assertions(2);

        const ErrorComponent = jest.fn(function TestError() {
            return <span>Oh dear</span>;
        });

        const suggestionError = new Error("fail");

        const props = {
            itemToString: item => item,
            value: [],
            onChange: NOOP,
            getSuggestedItems: () => {
                throw suggestionError;
            },
            ErrorComponent
        };
        const wrapper = mountStable(<MultiPicker {...props }/>);

        await changeInputValueAndUpdate(wrapper, "some text");

        expect(ErrorComponent).toHaveBeenCalledWith({ error: suggestionError, inputValue: "some text" }, {});
        expect(wrapper.find(PickerSuggestions).getDOMNode()).toMatchSnapshot();
    });

    it("shows popover if 'itemToPopover' prop is supplied and returns content", () => {
        expect.assertions(4);

        const props = {
            itemToString: item => item,
            itemToPopover: jest.fn(() => <span>some popover</span>),
            value: ["some-item"],
            onChange: NOOP,
            getSuggestedItems: () => []
        };
        const wrapper = mountStable(<MultiPicker { ...props } />);
        expect(wrapper).toContainExactlyOneMatchingElement(Chip);
        const chipBeforePopover = wrapper.find(Chip);

        // open popover
        act(() => {
            chipBeforePopover.props().onMouseEnter({ currentTarget: chipBeforePopover.getDOMNode() });
        });
        wrapper.update();
        expect(props.itemToPopover).toHaveBeenCalledWith("some-item");
        expect(wrapper.find(Popover)).toHaveProp("open", true);

        // close popover
        act(() => {
            chipBeforePopover.props().onMouseLeave();
        });
        wrapper.update();
        expect(wrapper.find(Popover)).toHaveProp("open", false);
    });

    it("does not open popover if 'itemToPopover' prop returns nothing", () => {
        expect.assertions(4);

        const props = {
            itemToString: item => item,
            itemToPopover: jest.fn(() => false),
            value: ["some-item"],
            onChange: NOOP,
            getSuggestedItems: () => []
        };
        const wrapper = mountStable(<MultiPicker { ...props } />);
        expect(wrapper).toContainExactlyOneMatchingElement(Chip);
        const chipBeforePopover = wrapper.find(Chip);
        act(() => {
            chipBeforePopover.props().onMouseEnter({ currentTarget: chipBeforePopover.getDOMNode() });
        });
        wrapper.update();

        expect(props.itemToPopover).toHaveBeenCalledWith("some-item");
        expect(wrapper).toContainExactlyOneMatchingElement(Popover);
        expect(wrapper.find(Popover)).toHaveProp("open", false);

    });

    it("propagates 'required', 'helperText', 'name', 'autoFocus' and 'variant' fields to the TextField component", () => {
        expect.assertions(1);

        const baseProps = {
            itemToString: item => item,
            value: [],
            onChange: NOOP,
            getSuggestedItems: () => []
        };
        const propsToPropagate = {
            variant: "outlined",
            required: true,
            name: "some-picker",
            helperText: "Some Helper Text",
            autoFocus: true
        };
        const wrapper = mountStable(<MultiPicker { ...baseProps } { ...propsToPropagate } />);

        expect(wrapper.find(TextField)).toHaveProp(propsToPropagate);
    });

    it("uses 'outlined' style for the chips when the 'filled' variant is used", () => {
        expect.assertions(1);

        const props = {
            itemToString: item => item,
            value: ["some item"],
            onChange: NOOP,
            getSuggestedItems: () => [],
            variant: "filled"
        };
        const wrapper = mountStable(<MultiPicker { ...props } />);

        expect(wrapper.find(Chip)).toHaveProp("variant", "outlined");
    });


    it("applies the 'maxDropdownHeight' prop to the dropdown", async () => {
        expect.assertions(1);

        const props = {
            itemToString: item => item,
            value: [],
            onChange: NOOP,
            getSuggestedItems: () => [],
            maxDropdownHeight: 100
        };
        const wrapper = mountStable(<MultiPicker { ...props } />);
        await changeInputValueAndUpdate(wrapper, "some text");

        expect(wrapper.find(Paper)).toHaveStyle("maxHeight", 100);
    });

    it("shows the focus styling on the picker when it has the focus", async () => {
        expect.assertions(1);

        const props = {
            itemToString: item => item,
            value: [],
            onChange: NOOP,
            getSuggestedItems: () => []
        };
        const wrapper = mountStable(<MultiPicker { ...props } label="some label" />);
        wrapper.find("input").simulate("focus", { target: wrapper.find("input") });
        await delay(100);
        wrapper.update();
        expect(wrapper.find(InputLabel).find("label")).toHaveProp("className", stringContaining("MuiFormLabel-focused"));
    });

    it("shows the dropdown when the picker is focused (if the `showDropdownOnFocus` prop is set)", async () => {
        expect.assertions(1);

        const props = {
            itemToString: item => item,
            value: [],
            onChange: NOOP,
            getSuggestedItems: () => ["some item"],
            showDropdownOnFocus: true
        };
        const wrapper = mountStable(<MultiPicker { ...props } />);
        wrapper.find("input").simulate("focus");
        await delay(100);
        wrapper.update();
        expect(wrapper).toContainExactlyOneMatchingElement(PickerSuggestions);
    });

});
