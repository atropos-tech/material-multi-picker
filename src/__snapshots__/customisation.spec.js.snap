// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiPicker component renders custom chip avatars 1`] = `
<WithStyles(Avatar)
  childrenClassName="MuiChip-avatarChildren"
  className="MuiChip-avatar"
  src="./missing-image"
>
  <Avatar
    childrenClassName="MuiChip-avatarChildren"
    className="MuiChip-avatar"
    classes={
      Object {
        "colorDefault": "MuiAvatar-colorDefault",
        "img": "MuiAvatar-img",
        "root": "MuiAvatar-root",
      }
    }
    component="div"
    src="./missing-image"
  >
    <div
      className="MuiAvatar-root MuiChip-avatar"
    >
      <img
        className="MuiAvatar-img"
        src="./missing-image"
      />
    </div>
  </Avatar>
</WithStyles(Avatar)>
`;

exports[`MultiPicker component renders custom chip labels 1`] = `
<WithStyles(Chip)
  className="MultiPicker-pickerChip"
  label="some label"
  onDelete={[Function]}
  tabIndex={-1}
  variant="default"
>
  <Chip
    className="MultiPicker-pickerChip"
    classes={
      Object {
        "avatar": "MuiChip-avatar",
        "avatarChildren": "MuiChip-avatarChildren",
        "avatarColorPrimary": "MuiChip-avatarColorPrimary",
        "avatarColorSecondary": "MuiChip-avatarColorSecondary",
        "clickable": "MuiChip-clickable",
        "clickableColorPrimary": "MuiChip-clickableColorPrimary",
        "clickableColorSecondary": "MuiChip-clickableColorSecondary",
        "colorPrimary": "MuiChip-colorPrimary",
        "colorSecondary": "MuiChip-colorSecondary",
        "deletable": "MuiChip-deletable",
        "deletableColorPrimary": "MuiChip-deletableColorPrimary",
        "deletableColorSecondary": "MuiChip-deletableColorSecondary",
        "deleteIcon": "MuiChip-deleteIcon",
        "deleteIconColorPrimary": "MuiChip-deleteIconColorPrimary",
        "deleteIconColorSecondary": "MuiChip-deleteIconColorSecondary",
        "deleteIconOutlinedColorPrimary": "MuiChip-deleteIconOutlinedColorPrimary",
        "deleteIconOutlinedColorSecondary": "MuiChip-deleteIconOutlinedColorSecondary",
        "icon": "MuiChip-icon",
        "iconColorPrimary": "MuiChip-iconColorPrimary",
        "iconColorSecondary": "MuiChip-iconColorSecondary",
        "label": "MuiChip-label",
        "outlined": "MuiChip-outlined",
        "outlinedPrimary": "MuiChip-outlinedPrimary",
        "outlinedSecondary": "MuiChip-outlinedSecondary",
        "root": "MuiChip-root",
      }
    }
    color="default"
    component="div"
    label="some label"
    onDelete={[Function]}
    tabIndex={-1}
    variant="default"
  >
    <div
      className="MuiChip-root MuiChip-deletable MultiPicker-pickerChip"
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      role="button"
      tabIndex={-1}
    >
      <span
        className="MuiChip-label"
      >
        some label
      </span>
      <pure(Cancel)
        className="MuiChip-deleteIcon"
        onClick={[Function]}
      >
        <Cancel
          className="MuiChip-deleteIcon"
          onClick={[Function]}
        >
          <WithStyles(SvgIcon)
            className="MuiChip-deleteIcon"
            onClick={[Function]}
          >
            <SvgIcon
              className="MuiChip-deleteIcon"
              classes={
                Object {
                  "colorAction": "MuiSvgIcon-colorAction",
                  "colorDisabled": "MuiSvgIcon-colorDisabled",
                  "colorError": "MuiSvgIcon-colorError",
                  "colorPrimary": "MuiSvgIcon-colorPrimary",
                  "colorSecondary": "MuiSvgIcon-colorSecondary",
                  "fontSizeInherit": "MuiSvgIcon-fontSizeInherit",
                  "fontSizeLarge": "MuiSvgIcon-fontSizeLarge",
                  "fontSizeSmall": "MuiSvgIcon-fontSizeSmall",
                  "root": "MuiSvgIcon-root",
                }
              }
              color="inherit"
              component="svg"
              fontSize="default"
              onClick={[Function]}
              viewBox="0 0 24 24"
            >
              <svg
                aria-hidden="true"
                className="MuiSvgIcon-root MuiChip-deleteIcon"
                focusable="false"
                onClick={[Function]}
                role="presentation"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
                />
              </svg>
            </SvgIcon>
          </WithStyles(SvgIcon)>
        </Cancel>
      </pure(Cancel)>
    </div>
  </Chip>
</WithStyles(Chip)>
`;

exports[`MultiPicker component renders custom suggestion components 1`] = `
<WithStyles(PickerSuggestions)
  SuggestionComponent={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "inputValue": "some text",
            "isHighlighted": false,
            "item": "some suggestion",
            "itemId": "some suggestion",
          },
          Object {},
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": <span>
            Some Custom Suggestion
          </span>,
        },
      ],
    }
  }
  clearItems={[Function]}
  clearSelection={[Function]}
  closeMenu={[Function]}
  getItemProps={[Function]}
  getLabelProps={[Function]}
  getMenuProps={[Function]}
  getRootProps={[Function]}
  getSuggestedItems={[Function]}
  getToggleButtonProps={[Function]}
  highlightedIndex={null}
  id="downshift-0"
  inputValue="some text"
  itemToString={[Function]}
  openMenu={[Function]}
  pickedItems={Array []}
  reset={[Function]}
  selectHighlightedItem={[Function]}
  selectItem={[Function]}
  selectItemAtIndex={[Function]}
  selectedItem={null}
  setHighlightedIndex={[Function]}
  setItemCount={[Function]}
  setState={[Function]}
  suggestions={
    Array [
      "some suggestion",
    ]
  }
  toggleMenu={[Function]}
  unsetItemCount={[Function]}
>
  <PickerSuggestions
    SuggestionComponent={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "inputValue": "some text",
              "isHighlighted": false,
              "item": "some suggestion",
              "itemId": "some suggestion",
            },
            Object {},
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": <span>
              Some Custom Suggestion
            </span>,
          },
        ],
      }
    }
    classes={
      Object {
        "highlighted": "PickerSuggestions-highlighted",
      }
    }
    clearItems={[Function]}
    clearSelection={[Function]}
    closeMenu={[Function]}
    getItemProps={[Function]}
    getLabelProps={[Function]}
    getMenuProps={[Function]}
    getRootProps={[Function]}
    getSuggestedItems={[Function]}
    getToggleButtonProps={[Function]}
    highlightedIndex={null}
    id="downshift-0"
    inputValue="some text"
    itemToString={[Function]}
    openMenu={[Function]}
    pickedItems={Array []}
    reset={[Function]}
    selectHighlightedItem={[Function]}
    selectItem={[Function]}
    selectItemAtIndex={[Function]}
    selectedItem={null}
    setHighlightedIndex={[Function]}
    setItemCount={[Function]}
    setState={[Function]}
    suggestions={
      Array [
        "some suggestion",
      ]
    }
    toggleMenu={[Function]}
    unsetItemCount={[Function]}
  >
    <WithStyles(MenuItem)
      aria-selected={false}
      className=""
      id="downshift-0-item-0"
      key="some suggestion"
      onClick={[Function]}
      onMouseDown={[Function]}
      onMouseMove={[Function]}
      role="menuitem"
      style={
        Object {
          "height": "auto",
          "padding": 0,
        }
      }
    >
      <MenuItem
        aria-selected={false}
        className=""
        classes={
          Object {
            "gutters": "MuiMenuItem-gutters",
            "root": "MuiMenuItem-root",
            "selected": "MuiMenuItem-selected",
          }
        }
        component="li"
        disableGutters={false}
        id="downshift-0-item-0"
        onClick={[Function]}
        onMouseDown={[Function]}
        onMouseMove={[Function]}
        role="menuitem"
        style={
          Object {
            "height": "auto",
            "padding": 0,
          }
        }
      >
        <WithStyles(ListItem)
          aria-selected={false}
          button={true}
          className="MuiMenuItem-root MuiMenuItem-gutters"
          component="li"
          disableGutters={false}
          id="downshift-0-item-0"
          onClick={[Function]}
          onMouseDown={[Function]}
          onMouseMove={[Function]}
          role="menuitem"
          style={
            Object {
              "height": "auto",
              "padding": 0,
            }
          }
          tabIndex={-1}
        >
          <ListItem
            ContainerComponent="li"
            alignItems="center"
            aria-selected={false}
            button={true}
            className="MuiMenuItem-root MuiMenuItem-gutters"
            classes={
              Object {
                "alignItemsFlexStart": "MuiListItem-alignItemsFlexStart",
                "button": "MuiListItem-button",
                "container": "MuiListItem-container",
                "default": "MuiListItem-default",
                "dense": "MuiListItem-dense",
                "disabled": "MuiListItem-disabled",
                "divider": "MuiListItem-divider",
                "focusVisible": "MuiListItem-focusVisible",
                "gutters": "MuiListItem-gutters",
                "root": "MuiListItem-root",
                "secondaryAction": "MuiListItem-secondaryAction",
                "selected": "MuiListItem-selected",
              }
            }
            component="li"
            dense={false}
            disableGutters={false}
            disabled={false}
            divider={false}
            id="downshift-0-item-0"
            onClick={[Function]}
            onMouseDown={[Function]}
            onMouseMove={[Function]}
            role="menuitem"
            selected={false}
            style={
              Object {
                "height": "auto",
                "padding": 0,
              }
            }
            tabIndex={-1}
          >
            <MergeListContext
              alignItems="center"
              dense={false}
            >
              <WithStyles(ButtonBase)
                aria-selected={false}
                className="MuiListItem-root MuiListItem-default MuiListItem-gutters MuiListItem-button MuiMenuItem-root MuiMenuItem-gutters"
                component="li"
                disabled={false}
                focusVisibleClassName="MuiListItem-focusVisible"
                id="downshift-0-item-0"
                onClick={[Function]}
                onMouseDown={[Function]}
                onMouseMove={[Function]}
                role="menuitem"
                style={
                  Object {
                    "height": "auto",
                    "padding": 0,
                  }
                }
                tabIndex={-1}
              >
                <ButtonBase
                  aria-selected={false}
                  centerRipple={false}
                  className="MuiListItem-root MuiListItem-default MuiListItem-gutters MuiListItem-button MuiMenuItem-root MuiMenuItem-gutters"
                  classes={
                    Object {
                      "disabled": "MuiButtonBase-disabled",
                      "focusVisible": "MuiButtonBase-focusVisible",
                      "root": "MuiButtonBase-root",
                    }
                  }
                  component="li"
                  disableRipple={false}
                  disableTouchRipple={false}
                  disabled={false}
                  focusRipple={false}
                  focusVisibleClassName="MuiListItem-focusVisible"
                  id="downshift-0-item-0"
                  onClick={[Function]}
                  onMouseDown={[Function]}
                  onMouseMove={[Function]}
                  role="menuitem"
                  style={
                    Object {
                      "height": "auto",
                      "padding": 0,
                    }
                  }
                  tabIndex={-1}
                  type="button"
                >
                  <li
                    aria-selected={false}
                    className="MuiButtonBase-root MuiListItem-root MuiListItem-default MuiListItem-gutters MuiListItem-button MuiMenuItem-root MuiMenuItem-gutters"
                    id="downshift-0-item-0"
                    onBlur={[Function]}
                    onClick={[Function]}
                    onContextMenu={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseLeave={[Function]}
                    onMouseMove={[Function]}
                    onMouseUp={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    role="menuitem"
                    style={
                      Object {
                        "height": "auto",
                        "padding": 0,
                      }
                    }
                    tabIndex={-1}
                  >
                    <mockConstructor
                      inputValue="some text"
                      isHighlighted={false}
                      item="some suggestion"
                      itemId="some suggestion"
                      key=".0"
                    >
                      <span>
                        Some Custom Suggestion
                      </span>
                    </mockConstructor>
                    <NoSsr
                      defer={false}
                      fallback={null}
                    >
                      <WithStyles(TouchRipple)
                        center={false}
                        innerRef={[Function]}
                      >
                        <TouchRipple
                          center={false}
                          classes={
                            Object {
                              "child": "MuiTouchRipple-child",
                              "childLeaving": "MuiTouchRipple-childLeaving",
                              "childPulsate": "MuiTouchRipple-childPulsate",
                              "ripple": "MuiTouchRipple-ripple",
                              "ripplePulsate": "MuiTouchRipple-ripplePulsate",
                              "rippleVisible": "MuiTouchRipple-rippleVisible",
                              "root": "MuiTouchRipple-root",
                            }
                          }
                        >
                          <TransitionGroup
                            childFactory={[Function]}
                            className="MuiTouchRipple-root"
                            component="span"
                            enter={true}
                            exit={true}
                          >
                            <span
                              className="MuiTouchRipple-root"
                            />
                          </TransitionGroup>
                        </TouchRipple>
                      </WithStyles(TouchRipple)>
                    </NoSsr>
                  </li>
                </ButtonBase>
              </WithStyles(ButtonBase)>
            </MergeListContext>
          </ListItem>
        </WithStyles(ListItem)>
      </MenuItem>
    </WithStyles(MenuItem)>
  </PickerSuggestions>
</WithStyles(PickerSuggestions)>
`;

exports[`MultiPicker component shows an custom error message if the user provides the ErrorComponent prop 1`] = `
<WithStyles(PickerSuggestions)
  ErrorComponent={
    [MockFunction] {
      "calls": Array [
        Array [
          Object {
            "error": [Error: fail],
            "inputValue": "some text",
          },
          Object {},
        ],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": <span>
            Oh dear
          </span>,
        },
      ],
    }
  }
  clearItems={[Function]}
  clearSelection={[Function]}
  closeMenu={[Function]}
  getItemProps={[Function]}
  getLabelProps={[Function]}
  getMenuProps={[Function]}
  getRootProps={[Function]}
  getSuggestedItems={[Function]}
  getToggleButtonProps={[Function]}
  highlightedIndex={null}
  id="downshift-0"
  inputValue="some text"
  itemToString={[Function]}
  openMenu={[Function]}
  pickedItems={Array []}
  reset={[Function]}
  selectHighlightedItem={[Function]}
  selectItem={[Function]}
  selectItemAtIndex={[Function]}
  selectedItem={null}
  setHighlightedIndex={[Function]}
  setItemCount={[Function]}
  setState={[Function]}
  suggestions={[Error: fail]}
  toggleMenu={[Function]}
  unsetItemCount={[Function]}
>
  <PickerSuggestions
    ErrorComponent={
      [MockFunction] {
        "calls": Array [
          Array [
            Object {
              "error": [Error: fail],
              "inputValue": "some text",
            },
            Object {},
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": <span>
              Oh dear
            </span>,
          },
        ],
      }
    }
    classes={
      Object {
        "highlighted": "PickerSuggestions-highlighted",
      }
    }
    clearItems={[Function]}
    clearSelection={[Function]}
    closeMenu={[Function]}
    getItemProps={[Function]}
    getLabelProps={[Function]}
    getMenuProps={[Function]}
    getRootProps={[Function]}
    getSuggestedItems={[Function]}
    getToggleButtonProps={[Function]}
    highlightedIndex={null}
    id="downshift-0"
    inputValue="some text"
    itemToString={[Function]}
    openMenu={[Function]}
    pickedItems={Array []}
    reset={[Function]}
    selectHighlightedItem={[Function]}
    selectItem={[Function]}
    selectItemAtIndex={[Function]}
    selectedItem={null}
    setHighlightedIndex={[Function]}
    setItemCount={[Function]}
    setState={[Function]}
    suggestions={[Error: fail]}
    toggleMenu={[Function]}
    unsetItemCount={[Function]}
  >
    <mockConstructor
      error={[Error: fail]}
      inputValue="some text"
    >
      <span>
        Oh dear
      </span>
    </mockConstructor>
  </PickerSuggestions>
</WithStyles(PickerSuggestions)>
`;
